dimensionless factory methods renamed

philwalk opened the pull request:
Addresses comment about unsupported cases in #65.

It was necessary to rename dimensionless factory methods from `def apply` to `def mat`.
In other words, when dimensions are declared, the factory name is `def apply[M <: Int, N <: Int]` as before,
and otherwise it's `def mat(...)`.

The name collisions are a side-effect of supporting varargs of tuples, although it might have been inevitable anyway.  Any method that takes a single varargs sequence has the same signature after erasure, even if one has varargs `Double *` and the other has `Tuple *`.

Here's an example of the new method names,  from `MatExtensionsTest`:
```
  test("Mat from Tuple literals match whether dimensions provided") {
    val m01: Mat[3, 2] = Mat(((11, 21), (31, 41), (51, 61)))
    val m02            =  mat((11, 21), (31, 41), (51, 61))
    assert(m01.strictEquals(m02))
  }
```

There is still a need to verify good test coverage.  Also, there are some unused code in `Mat.scala` that can be cleaned up, but I'll have to get to it later.   I need to push this code so people can start looking at it.

Feedback is welcome.

dragonfly-ai commented:
Please forgive my unresponsiveness.  I hope to return to productivity next week.

dragonfly-ai commented:
I wonder if we should just embrace:

```val m:Mat[3,3] = Mat[3,3](
  1.0, 2.0, 3.0,
  4.0, 5.0, 6.0,
  7.0, 8.0, 9.0
)
```
and
```val mft:Mat[3,3] = Mat.fromTuples[3,3](
  (1.0, 2.0, 3.0),
  (4.0, 5.0, 6.0),
  (7.0, 8.0, 9.0)
)
```
and
```val mft:Mat[3,3] = Mat.fromTuple[3,3](
  (
    (1.0, 2.0, 3.0),
    (4.0, 5.0, 6.0),
    (7.0, 8.0, 9.0)
  )
)
```

philwalk commented:
I'm experimenting with some techniques to avoid ambiguous overload errors.   I'll update you on my progress before too long.  Otherwise, your suggestions make sense.

dragonfly-ai commented:
This PR has some cool things in it but before the merge, I have some questions:

```scala
type Number = Int | Float | Long | Double
```
I guess this helps to interpret tuples, but are you sure we need it?  If we need it for `Mat` why didn't we need it for `Vec`?

```scala
  def apply[M <: Int, N <: Int](val0: Number, values: Number*)(using ValueOf[M], ValueOf[N]):Mat[M, N] = {
```
Why change the single parameter from `values: Double*` to `val0: Number, values: Number*`?  Is it about precluding 1x1 matrices?  If not, why?

```scala
    new Mat[M, N](NArray[Double]((val0 :: values.toList).map(toDouble) *))
```

Later in that same `apply` method: `new Mat[M, N](NArray[Double]((val0 :: values.toList).map(toDouble) *))`.  We can do this in a faster and lighter way.  Let's not create and convert between 4 data structures to do the work of a single loop.

```scala
/** Construct 1xN a matrix from a varargs of type Numeric.
   * @param values One-dimensional array of Numeric.
   */
  def mat(values: Number *): Mat[? <: Int, ? <: Int] = {
    val rows: Int = 1
```
If it always returns a `Mat[1,N]`, why not change the signature to: `def mat(values: Number *): Mat[1, ? <: Int] = {` or better still: `def columnMat(values: Number *): Mat[1, ? <: Int] = {`.

Relatedly, did you see [these extension methods?](https://github.com/dragonfly-ai/slash/blob/main/slash/shared/src/main/scala/slash/matrix/package.scala#L29)

```scala
  extension[N <: Int] (thisVector: Vec[N])(using ValueOf[N]) {
    inline def asRowMatrix: Mat[1, N] = Mat[1, N](thisVector.asInstanceOf[NArray[Double]])
    inline def asColumnMatrix: Mat[N, 1] = Mat[N, 1](thisVector.asInstanceOf[NArray[Double]])
```
I'm also hesitant about any method that returns `Vec[? <: Int]` or `Mat[? <: Int, ? <: Int]` because these `? <: Int` type parameters undermine a core property of the library.

Can you tell me why users might need vector and matrix dimensions that have no runtime guarantees whatsoever?

SLASH `Vec` and `Mat` already support constant dimensions and dimensions defined by a runtime variable.  I think that can cover all cases.  Have I missed any?

Thank you for your contributions and patience!  I appreciate your pull requests and I will review them all at some speed or another.  :)

philwalk commented:
> If we need it for `Mat` why didn't we need it for `Vec`?
the only reason for bringing in the `Number` type is to allow expressions with `Int` literals or other numeric values, although perhaps we don't need it, I'll investigate.

> Why change the single parameter from `values: Double*` to `val0: Number, values: Number*`
No longer necessary, now that we renamed some conflicting `apply` methods to `fromTuple`, so I'll revert this.

For context, it was a way to differentiate `apply` method signatures. After erasure, methods that accept varargs are indistinguishable, since they look like this at runtime: `def apply(args: Seq[?]) ...`.

As an aside, there is no literal expression for a `Tuple1` in scala3 at present (there is some discussion on the topic [here](https://contributors.scala-lang.org/t/syntax-for-type-tuple-with-one-element/6974/34)).


> If it always returns a `Mat[1,N]`, why not change the signature to: `def mat(values: Number *): Mat[1, ? <: Int] = {`

That seems to be debris from an early experiment, I'll remove it.


BTW, I suspect you're aware of it, but I occasionally see this error, although it's intermittent.
```
==> X VectorBoundsTest. VectorBounds.contains  0.003s munit.FailException: tests/shared/src/test/scala/VectorBoundsTest.scala:54 assertion failed
53:      v.normalize()
54:      assert(!bounds.contains(v * (bounds.boundingRadius + r.between(0.00000001, 1.0))))
55:      i = i + 1
```

dragonfly-ai commented:
Looking better, for sure.  I'm mainly holding back because of this:

```scala
transparent inline def apply[T <: Tuple](inline t: T *) = { ... }
```

How would users make use of this factory if the `Mat` it returns has no compile time dimensional compatibility with any other `Mat` or `Vec`?

Do you have a use in mind?

philwalk commented:
> How would users make use of this factory if the `Mat` it returns has no compile time dimensional compatibility with any other `Mat` or `Vec`

It's invaluable that this library lets me debug my equations using compile-time dimensioned Mats and Vecs, created from literal expressions.   But I mostly I process weekly financial data, and the dimensions of the matrices aren't fixed.  I could update my source code each week, but after debugging my math, runtime checks are good enough.  Dimensions currently range up to several thousand rows by 200+ columns.

Although matrices with no compile time dimensions are less convenient to use, the code below shows some possibilities.
Another approach would be to regenerate code each week after acquiring necessary dimensions.

Some convenient definitions for the sake of simplified exposition:
`Mat` :  has full compile-time dimensions
`LMat`: only the number of rows known at compile time
`RMat`: only the number of columns known at compile time
`NMat`: neither dimension is known at compile time

`def timesA(...)` multiplies a `Mat` by a `NMat`, returning an `LMat`

`def timesB(...)` multiplies two `NMat`s returning an `NMat`

```scala
import slash.matrix.Mat
import scala.compiletime.ops.int.*

val matA = Mat[3,2](    ((11, 21), (31, 41), (51, 61)))
val matB = Mat.fromTuple((11, 21), (31, 41), (51, 61)).transpose

printf("%s\n", timesA(matA, matB))

// multiple a `Mat` times an `LMat`
def timesA[M <: Int,N <: Int](left: Mat[M,N], right: Mat[?,?])(using ValueOf[M], ValueOf[N]): Mat[M,?] = {
  require(valueOf[N] == right.rows)
  type O = right.columns.type
  left * right.asInstanceOf[Mat[N,O]]
}

val matC = Mat.fromTuple((11, 21), (31, 41), (51, 61))
printf("%s\n", timesB(matC, matB))

// multiply two `NMat`s
def timesB(left: Mat[?,?], right: Mat[?,?]) = {
  require(left.columns == right.rows)
  type M = left.rows.type
  type N = left.columns.type
  type O = right.columns.type
  left.asInstanceOf[Mat[M,N]] * right.asInstanceOf[Mat[N,O]]
}
```

When translating `python`, `breeze` or `R` projects to `slash`, the most important requirement is flexibility and ease of translation.

dragonfly-ai commented:
[`VectorSpace`](https://github.com/dragonfly-ai/slash/blob/main/slash/shared/src/main/scala/slash/vector/VectorSpace.scala) and [`MatrixSpace`](https://github.com/dragonfly-ai/slash/blob/main/slash/shared/src/main/scala/slash/matrix/MatrixSpace.scala) exist to address this issue.

Are they falling short somehow?

```scala
class MatrixSpace[M0 <: Int, N0 <: Int](val rowVectorSpace:VectorSpace[M0], val columnVectorSpace:VectorSpace[N0]) {

  val rowDimension:Int = rowVectorSpace.dimension
  val columnDimension:Int = columnVectorSpace.dimension

  opaque type M <: Int = rowVectorSpace.N
  opaque type N <: Int = columnVectorSpace.N
  opaque type MN <: Int = M * N
...
}
```

Using this, you can create a new mat:
```scala
val ms = MatrixSpace(rows, columns)

val m: Mat[ms.M, ms.N] = Mat.fromTuple[ms.M, ms.N]( ... )
```

I realize that it might be hard to know the dimensions before knowing a matrix, so maybe we should add a `matSpace` method to the `Mat` type and create a directory of existing `MatrixSpace` instances in the `Mat` singleton.

I created SLASH because I also want to process weekly financial data, so I certainly want to help you do it!

Please lmk if I can elaborate on any of this.  I want to hear your thoughts.

dragonfly-ai commented:
Maybe a read from tuple literal of unknown dimensions could return a `Mat` and its corresponding `MatrixSpace`?

philwalk commented:
The design is starting to make sense to me.   It would be useful to have the MatrixSpace objects managed on behalf of developers to the extent possible.

> maybe we should add a matSpace method to the Mat type and create a directory of existing MatrixSpace instances in the Mat singleton.

That sounds good.   I'll investigate how this PR could be updated w.r.t. this idea.

dragonfly-ai commented:
Actually, the singleton idea won't really work.  It provides a runtime solution to a compile time problem.

dragonfly-ai commented:
`VectorSpace` and `MatrixSpace` are meant to wrap dimension types that can be shared across matrices and vectors.

This is a way to provide compile time certainty about dimensions defined at runtime, but I admit it's a little clunky.

Please lmk if I can explain anything better or if you'd like to see more examples.

I'm particularly interested in knowing whether or not they can help in your use cases.  If they fall short, it could mean that we have to rethink the whole design of SLASH.

dragonfly-ai commented:
> MatrixSpace objects managed on behalf of developers to the extent possible.

Agreed.  I'd love to hear your ideas about how to do that.

philwalk commented:
I've been working on this, it's quite subtle as you pointed out.  I might have something soon ...

philwalk commented:
Here an update on my thinking, and some conclusions I've reached.
TLDR; We'll have to require type parameters to `Mat.fromTuple[M <: Int, N <: Int, T <: Tuple](t: T)`

The `Mat.fromTuple(tuplerows: T)...` method should in principle be able to return a perfectly good `Mat[M,N]` object without having to specify the dimensions.   As currently implemented, however, it returns a union type "Mat[3,5] | Mat[5,3]".  The values of `rows` and `columns` is correct, but the compile time type parameters M and N are unions.

The problem stems from a confusing aspect of scala3 tuple literals.
At the type level, these are unambiguous:

- `Tuple2[Tuple3[Int,Int,Int],Tuple3[Int,Int,Int]]`
- `Tuple1[Int,Int,Int]`
- `Tuple1[Tuple3[Int,Int,Int]]`


The first type corresponds to `((1,2,3), (4,3,5))` which is  `Tuple2[Tuple3[Int,Int,Int],Tuple3[Int,Int,Int]]`
Delete a row and there's a temptation to think of it as a `Tuple1[Tuple3[Int,Int,Int]]` but it's actually a `Tuple1[Int,Int,Int]` inside parens.

So we need to add type parameters for the dimensions.

On the topic of simplified `VectorSpace` capability,  my idea is to preallocate dimensions 0 to 1999 so the usage is maximally similar to using Int-literal dimensions.  The following demo script shows how it would work.
I did a `publishLocal` of my working code with version `0.3.96` for testing purposes.

The extension method at the bottom allows us to cleanly convert things to 'conventional' format, so we could read from a csv file and then convert it.  The only advantage of applying the dimensions after the `Mat` is created is it enables doing things before you know what the dimensions are (like showing the dimensions).

```scala
#!/usr/bin/env -S scala-cli shebang --scratch-dir c:/tmp/jars 
//package vast.apps

//> using scala 3.6.4
//> using dep ai.dragonfly::slash:0.3.96

import slash.matrix.*
import slash.matrix.Mat.*
import slash.matrix.DimSpace.D
import slash.vector.*
import narr.*

import scala.compiletime.ops.int.*

object SlashSpace {
  def main(args: Array[String]): Unit = {
   
    val rows = 2
    val columns = 3

    val ms = MatrixSpace(rows, columns)
    val m0: Mat[ms.M, ms.N] = ms.random
    val t0: Mat[ms.N, ms.M] = m0.transpose
    printf("%s\n", m0)
    printf("%s\n", t0)
    //printf("%s\n", m0 * t0) // compile error
   
    // change VectorSpace dimensioned mat to dimensions [2,3] 
    val rMat: Mat[2,3] = m0.as[2,3]
    printf("%s\n", rMat)
    printf("%s\n", rMat.transpose * rMat)

    import slash.matrix.DimSpace.D.*
    type M = _2.N
    type N = _3.N
    val m1 = Mat[M,N](1,2,3,4,5,6)
    val t1:  Mat[N,M] = m1.transpose
    printf("%s\n", m1)
    printf("%s\n", t1)
    printf("%s\n", m1 * t1)

    // change VectorSpace dimensioned mat to dimensions [2,3] 
    val m2: Mat[2,3] = m1.as[2,3]
    printf("%s\n", m2)
  }

  extension(m: Mat[?,?]) {
    def as[M <: Int, N <: Int](using ValueOf[M], ValueOf[N]): Mat[M,N] = Mat[M,N](m.values)
  }
}
```

dragonfly-ai commented:
I think every `Mat` constructor and factory needs the type parameters, but the code you wrote still serves a purpose to verify them against the inline tuple data.

I'm glad you saw the need for this:

```scala
def as[M <: Int, N <: Int](using ValueOf[M], ValueOf[N]): Mat[M,N] = Mat[M,N](m.values)
```
it could also be written:
  
```scala
def as[M <: Int, N <: Int](using ValueOf[M], ValueOf[N]): Mat[M,N] = {
  dimensionCheck(m.values.size, valueOf[M] * valueOf[N])
  m.asInstanceOf[Mat[M,N]]
}
```

I thought we had something like that already in `MatrixSpace` but clearly no.  What's the right name for it: `cast`, `assimilate`, `recognize`, `from`, `endorse`, `adopt`, `verify`?

Maybe I didn't include that method because I couldn't find a name, or maybe it's because if the user has a Matrix, she should already know its dimensions, so why cast?

"The only advantage of applying the dimensions after the Mat is created is it enables doing things before you know what the dimensions are (like showing the dimensions)."

This is the crux of the issue: what came first, the chicken (Matrix) or the egg (MatrixDimensions)?
We have to load the matrix to find out its dimensions, but we need its dimensions to allocate a matrix.

Unless ...

What about an intermediate data structure like:  `MatrixData` or `MatDat`?  If it wraps the dimensions, and the `NArray[Double]` values, one could use it to construct a couple of `VectorSpace`s, and a `MatrixSpace`, then create the matrix from the values.

Maybe `Vec` and `Mat` IO should live in the `VectorSpace`s, and `MatrixSpace` types.

Thank you for engaging!  You're making SLASH much better.

philwalk commented:
I added dimensionCheck and renamed `def as[M,N]` extension to `cast`, as you suggested.
Also, it isn't obvious if the preallocated `VecSpace` equivalents are useful for anything, since you need to know the dimensions before you can use them, so you might as well use `Int` literals instead, so you don't need to cast to `Int` literal dimensions at some later time.   I included them anyway since there might be a need that I'm not seeing.

As usual, feedback is welcome.

philwalk commented:
The compile error on the new source file `slash/shared/src/main/scala/slash/matrix/DimSpace.scala` says that it doesn't have a header, which I interpret to mean a Copyright Header, which it does have.  I saw this error before, but I can't remember what actually caused it.

dragonfly-ai commented:
> my idea is to preallocate dimensions 0 to 1999 so the usage is maximally similar to using Int-literal dimensions.

I saw that you added this.  What issue does it solve?

I can't undersand how `Vec[_199]` can ever help more than `Vec[199]`.

What have I missed?

philwalk commented:
> What have I missed?

I came to the same conclusion, but kept it so you could review it.
The underlying `Dm` class was inspired by `VectorSpace`.   The advantage of `Mat[_199]` over a `Mat[ms.M,ms.N]` is that it's a fully functional `Mat` object, whereas the `Mat` created by `MatrixSpace` can't be multiplied by its' transpose.   If we could fix that, I would understand the purpose of `MatrixSpace`.

Here's the code that shows the difference:
```scala
  val rows = 2
  val columns = 3

  def main(args: Array[String]): Unit = {
    val ms = MatrixSpace(rows, columns)
    val m0: Mat[ms.M, ms.N] = ms.random
    val t0: Mat[ms.N, ms.M] = m0.transpose
    printf("%s\n", m0)
    printf("%s\n", t0)
    printf("%s\n", m0 * t0) // compile error

    import slash.matrix.DimSpace.D.*
    type M = _2.N
    type N = _3.N
    val m1 = Mat[M,N](1,2,3,4,5,6)
    val t1:  Mat[N,M] = m1.transpose
    printf("%s\n", m1)
    printf("%s\n", t1)
    printf("%s\n", m1 * t1)

    // change VectorSpace dimensioned mat to dimensions [2,3] 
    val m2: Mat[2,3] = m1.as[2,3]
    printf("%s\n", m2)
  }
```

dragonfly-ai commented:
Thank you!  You found a bug!

These lines are missing from matrix space:

```scala
  given m: ValueOf[M] = rowVectorSpace.n
  given n: ValueOf[N] = columnVectorSpace.n
```

They go like this:
```scala

class MatrixSpace[M0 <: Int, N0 <: Int](val rowVectorSpace:VectorSpace[M0], val columnVectorSpace:VectorSpace[N0]) {

  val rowDimension:Int = rowVectorSpace.dimension
  val columnDimension:Int = columnVectorSpace.dimension

  opaque type M <: Int = rowVectorSpace.N
  opaque type N <: Int = columnVectorSpace.N
  opaque type MN <: Int = M * N

  given m: ValueOf[M] = rowVectorSpace.n
  given n: ValueOf[N] = columnVectorSpace.n
  ...
}
```

dragonfly-ai commented:
> whereas the `Mat` created by `MatrixSpace` can't be multiplied by its' transpose. If we could fix that, I would understand the purpose of `MatrixSpace`.

I wrote a fix for this, and for the bug in the `VectorBounds` test that you found earlier.

Haven't pushed yet, because I'm trying to get matrix formatting working.  So close!

philwalk commented:
I removed unused code, including the `DimSpace` object, and squashed commits for this branch.

dragonfly-ai commented on slash/shared/src/main/scala/slash/matrix/package.scala:
```diff
@@ -128,4 +128,14 @@ package object matrix {
+   */
+  extension(m: Mat[?,?]) {
+    def cast[M <: Int, N <: Int](using ValueOf[M], ValueOf[N]): Mat[M,N] = {
+      dimensionCheck(m.values.size, valueOf[M] * valueOf[N])
```
What if:
```scala
val m: Mat[3, 2] = ...
m.cast[2, 3]
```

philwalk replied:
What's your preference?  There's a `reshape[2,3]` alternate way to do that in #78, so we could disallow it here.
Something like this:
```scala
    def cast[M <: Int, N <: Int](using ValueOf[M], ValueOf[N]): Mat[M,N] = {
      assert(valueOf[M] == m.rows && valueOf[N] == m.columns)
      dimensionCheck(m.values.size, valueOf[M] * valueOf[N])
      Mat[M,N](m.values)
    }

```

dragonfly-ai replied:
I like this version better.

Why are we casting from a `Mat` instance, though, instead of from the `Mat` object or a `MatrixSpace` instance?

philwalk replied:
Hopefully the squash won't complicate the merge too much.
